File: mcp/tests/layer_boundary_test.rs
//! Import-direction lint for mcp/src. `.claude/rules/architecture.md` lists the layers top-down
//! — `server/` + tools → `opplan/` → `args` → `pipeline` → `llm` — so a module may reach every
//! module BELOW it in that row and none above. Today's crossings are a frozen allowance.

use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

const ORDER: [&str; 5] = ["llm", "pipeline", "args", "opplan", "server"];

/// Shrink this list; never add to it. Each entry is "file → module" as the scan reports it.
const ALLOWANCE: [(&str, &str); 2] = [
    ("pipeline/mod.rs → args", "the runner takes EditParams/ScrapeParams; args sits above it in the row"),
    ("pipeline/run.rs → args", "same seam: run() builds argv from args::EditParams"),
];

fn rank(module: &str) -> Option<usize> {
    ORDER.iter().position(|m| *m == module)
}

fn collect(dir: &Path, out: &mut Vec<PathBuf>) {
    let mut entries: Vec<_> = fs::read_dir(dir).expect("readable dir").flatten().collect();
    entries.sort_by_key(|e| e.path());
    for entry in entries {
        let path = entry.path();
        if path.is_dir() {
            collect(&path, out);
        } else if path.extension().is_some_and(|e| e == "rs") {
            out.push(path);
        }
    }
}

/// The top-level module that owns `rel` (`args/mod.rs` and `confine.rs` → `args`, `confine`);
/// `lib.rs` and `main.rs` are the crate root and own nothing.
fn owner(rel: &str) -> Option<String> {
    let first = rel.split('/').next()?;
    let name = first.strip_suffix(".rs").unwrap_or(first);
    if rel.contains('/') || !matches!(name, "lib" | "main") { Some(name.to_string()) } else { None }
}

/// Every module named after `crate::` (or `super::` in a file directly under src/, where super
/// IS the crate) on one code line, including the `crate::{a, b}` brace form.
fn referenced(line: &str, top_level_file: bool) -> Vec<String> {
    let code = line.split("//").next().unwrap_or("");
    let mut out = Vec::new();
    let mut prefixes = vec!["crate::"];
    if top_level_file {
        prefixes.push("super::");
    }
    for prefix in prefixes {
        for (i, _) in code.match_indices(prefix) {
            let rest = &code[i + prefix.len()..];
            if let Some(inner) = rest.strip_prefix('{') {
                let group = inner.split('}').next().unwrap_or("");
                out.extend(group.split(',').filter_map(|s| ident(s.trim())));
            } else if let Some(name) = ident(rest) {
                out.push(name);
            }
        }
    }
    out
}

fn ident(s: &str) -> Option<String> {
    let name: String = s.chars().take_while(|c| c.is_alphanumeric() || *c == '_').collect();
    if name.is_empty() || name == "self" { None } else { Some(name) }
}

/// Every cross-module reference in `src` as `(from_file, from_module, to_module)`.
fn scan(src: &Path) -> Vec<(String, String, String)> {
    let mut files = Vec::new();
    collect(src, &mut files);
    let mut edges = Vec::new();
    for path in files {
        let rel = path.strip_prefix(src).unwrap().to_string_lossy().replace('\\', "/");
        let Some(from) = owner(&rel) else { continue };
        let top_level_file = !rel.contains('/');
        let text = fs::read_to_string(&path).expect("readable source");
        let mut seen = BTreeSet::new();
        for line in text.lines() {
            for to in referenced(line, top_level_file) {
                if to != from && seen.insert(to.clone()) {
                    edges.push((rel.clone(), from.clone(), to));
                }
            }
        }
    }
    edges
}

/// `(rightward, unranked)`: edges that climb the row, and module names outside it.
fn violations(edges: &[(String, String, String)]) -> (Vec<String>, BTreeSet<String>) {
    let mut rightward = Vec::new();
    let mut unranked = BTreeSet::new();
    for (file, from, to) in edges {
        match (rank(from), rank(to)) {
            (Some(a), Some(b)) if b > a => rightward.push(format!("{file} → {to}")),
            (Some(_), Some(_)) => {}
            (a, b) => {
                if a.is_none() { unranked.insert(from.clone()); }
                if b.is_none() { unranked.insert(to.clone()); }
            }
        }
    }
    (rightward, unranked)
}

fn src_dir() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("src")
}

#[test]
fn the_scan_saw_the_whole_tree() {
    let edges = scan(&src_dir());
    assert!(edges.len() > 30, "cross-module references were found: {}", edges.len());
    let has = |f: &str, t: &str| edges.iter().any(|(_, a, b)| a == f && b == t);
    assert!(has("server", "opplan"), "server → opplan is indexed");
    assert!(has("opplan", "args"), "opplan → args is indexed");
    assert!(has("server", "pipeline"), "the crate::{{a, b}} form is indexed");
}

#[test]
fn no_module_reaches_a_layer_above_it() {
    let (rightward, _) = violations(&scan(&src_dir()));
    let unlisted: Vec<_> =
        rightward.iter().filter(|e| !ALLOWANCE.iter().any(|(k, _)| k == e)).collect();
    assert!(unlisted.is_empty(), "new upward references; move the code down instead:\n  {unlisted:?}");
}

#[test]
fn every_allowance_is_still_needed() {
    let (rightward, _) = violations(&scan(&src_dir()));
    let stale: Vec<_> =
        ALLOWANCE.iter().filter(|(k, _)| !rightward.iter().any(|e| e == k)).collect();
    assert!(stale.is_empty(), "the violation is gone; delete its allowance: {stale:?}");
}

#[test]
fn modules_outside_the_documented_row_are_the_known_helpers() {
    let (_, unranked) = violations(&scan(&src_dir()));
    let known: BTreeSet<String> = [
        "config", "confine", "deliver", "imagesize", "layout", "llmtransport", "locate",
        "outcome", "registry",
    ]
    .into_iter()
    .map(String::from)
    .collect();
    assert_eq!(unranked, known, "a module joined or left the tree; place it in the row");
}

#[test]
fn an_injected_upward_reference_is_caught() {
    let root = std::env::temp_dir().join(format!("stencil-layers-{}", std::process::id()));
    let _ = fs::remove_dir_all(&root);
    for dir in ["args", "pipeline", "llm", "server"] {
        fs::create_dir_all(root.join(dir)).unwrap();
    }
    fs::write(root.join("lib.rs"), "pub mod args;\npub mod pipeline;\n").unwrap();
    fs::write(root.join("args/mod.rs"), "use crate::pipeline::Run; // crate::server\npub fn a() {}\n").unwrap();
    fs::write(root.join("pipeline/mod.rs"), "use crate::llm::f;\npub struct Run;\n").unwrap();
    fs::write(root.join("llm/mod.rs"), "use crate::server::S;\npub fn f() -> S { S }\n").unwrap();
    fs::write(root.join("server/mod.rs"), "use crate::{args, llm};\npub struct S;\n").unwrap();
    fs::write(root.join("locate.rs"), "use super::server::S;\n").unwrap();
    let (rightward, unranked) = violations(&scan(&root));
    fs::remove_dir_all(&root).unwrap();
    assert_eq!(rightward, ["llm/mod.rs → server"]);
    assert_eq!(unranked.into_iter().collect::<Vec<_>>(), ["locate"]);
}

File: browser-extension/tests/layerBoundary.test.js
// Import-direction lint for browser-extension/src (.claude/rules/architecture.md): lib/ → config/ →
// llm/ → background/ → content/ → popup, options, crop, devtools. A layer may import from
// everything to its left and nothing to its right; today's crossings are a frozen allowance.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdirSync, mkdtempSync, readdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { dirname, join, relative, resolve } from 'node:path';
import { tmpdir } from 'node:os';
import { fileURLToPath } from 'node:url';

const SRC = resolve(dirname(fileURLToPath(import.meta.url)), '../src');
const LAYERS = [['lib'], ['config'], ['llm'], ['background'], ['content'],
  ['popup', 'options', 'crop', 'devtools']];
const RANK = new Map(LAYERS.flatMap((names, i) => names.map((n) => [n, i])));
const TOP = new Set(LAYERS[LAYERS.length - 1]);

// Shrink this list; never add to it. Each entry is "file → target" as reported below.
const ALLOWANCE = new Map([
  ['lib/dust/flight.js → config/motion.json',
    'lib reads a config table; the rule puts config/ right of lib/ (browser has it leftmost). The'
    + ' cloud family funnels it through this one file, so dustGrain/dustCloud take TUNE from here'],
  ['lib/image/videoFrames.js → llm/chatController.js',
    'MAX_IMAGE_EDGE lives in llm/controller.js; it belongs in a lib/ constants module'],
]);

const IMPORT = /^\s*(?:import|export)\b[^'"]*?\bfrom\s*['"]([^'"]+)['"]|^\s*import\s*['"]([^'"]+)['"]|\bimport\(\s*['"]([^'"]+)['"]\s*\)/gm;

const walk = (dir, out = []) => {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const path = join(dir, entry.name);
    if (entry.isDirectory()) walk(path, out);
    else if (entry.name.endsWith('.js')) out.push(path);
  }
  return out;
};

// Every relative import in `root`, as { from, to, edge } with layer names and "a → b" text.
export const scanEdges = (root) => {
  const edges = [];
  for (const file of walk(root)) {
    const rel = relative(root, file);
    const source = readFileSync(file, 'utf8');
    for (const m of source.matchAll(IMPORT)) {
      const spec = m[1] ?? m[2] ?? m[3];
      if (!spec.startsWith('.')) continue;
      const target = relative(root, resolve(dirname(file), spec));
      const from = rel.split('/')[0], to = target.split('/')[0];
      if (from !== to) edges.push({ from, to, edge: `${rel} → ${target}` });
    }
  }
  return edges;
};

export const findViolations = (edges) => {
  const rightward = [], sibling = [], unranked = [];
  for (const { from, to, edge } of edges) {
    if (!RANK.has(from) || !RANK.has(to)) { unranked.push(edge); continue; }
    if (RANK.get(to) > RANK.get(from)) rightward.push(edge);
    else if (TOP.has(from) && TOP.has(to)) sibling.push(edge);
  }
  return { rightward, sibling, unranked };
};

const edges = scanEdges(SRC);
const { rightward, sibling, unranked } = findViolations(edges);

test('the scan saw the whole tree', () => {
  assert.ok(walk(SRC).length > 150, 'browser-extension/src was read');
  assert.ok(edges.length > 100, 'cross-directory imports were found');
  assert.ok(edges.some((e) => e.from === 'popup' && e.to === 'lib'), 'popup → lib is indexed');
  assert.ok(edges.some((e) => e.from === 'llm' && e.to === 'config'), 'llm → config is indexed');
});

test('every src/ directory that imports or is imported has a layer', () => {
  assert.deepEqual(unranked, []);
});

test('no file imports from a layer to its right', () => {
  const unlisted = rightward.filter((e) => !ALLOWANCE.has(e));
  assert.deepEqual(unlisted, [], 'a new right-pointing import; move the code left instead');
});

test('the top-level pages do not import each other', () => {
  assert.deepEqual(sibling, [], 'popup, options, crop and devtools share code via lib/');
});

test('every allowance is still needed', () => {
  const seen = new Set(rightward);
  const stale = [...ALLOWANCE.keys()].filter((e) => !seen.has(e));
  assert.deepEqual(stale, [], 'the violation is gone; delete its allowance');
});

test('manifest entry points live in their layer', () => {
  const manifest = JSON.parse(readFileSync(resolve(SRC, '../manifest.json'), 'utf8'));
  assert.match(manifest.background.service_worker, /^src\/background\//);
  for (const cs of manifest.content_scripts) {
    for (const js of cs.js) assert.match(js, /^src\/content\//);
  }
});

test('an injected right-pointing import is caught', () => {
  const root = mkdtempSync(join(tmpdir(), 'stencil-layers-'));
  try {
    for (const dir of ['lib', 'llm', 'popup', 'options']) mkdirSync(join(root, dir));
    writeFileSync(join(root, 'lib/a.js'), "import { x } from '../llm/b.js';\nexport const y = x;\n");
    writeFileSync(join(root, 'llm/b.js'), "export * from '../lib/a.js';\nexport const x = 1;\n");
    writeFileSync(join(root, 'popup/p.js'), "const m = await import('../options/o.js');\n");
    writeFileSync(join(root, 'options/o.js'), "import '../lib/a.js';\n");
    const found = findViolations(scanEdges(root));
    assert.deepEqual(found.rightward, ['lib/a.js → llm/b.js']);
    assert.deepEqual(found.sibling, ['popup/p.js → options/o.js']);
    assert.deepEqual(found.unranked, []);
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
});
